### PR TITLE
feat: add heco and bsc mmr generation

### PR DESCRIPTION
**Pull request by songtianyi**
This pr include:
- code format
- some lint error fix
- heco and bsc mmr generation. Try to make as few modifications as possible to keep the original behavior

### CODE REVIEW

File: src/mmr/runner.rs [hunk 1]
```diff
@@ -51,26 +66,30 @@ impl Runner {
                 continue;
             }
 
-            if (ptr/30_000 * 30_000) as u64 + 30_000 > epoched && ffi::epoch(epoched) {
-                epoched += 30_000;
+            if (ptr / epoch_length * epoch_length) as u64 + epoch_length > epoched {
+                // notify epoch change
+                if &network == "Mainnet" {
+                    ffi::epoch(epoched);
```
Comment (wuminzhe) [inline src/mmr/runner.rs:72]: heco & bsc do not need to generate DAG files

Comment (songtianyi) [reply src/mmr/runner.rs:72]: Yes, the `Mainnet` represent ethereum mainnet

File: src/mmr/runner.rs [hunk 2]
```diff
@@ -31,15 +34,27 @@ impl Runner {
...
         // check network
         let network = self.check_network(client.as_ref()).await?;
-        let delay_blocks = if &network == "Mainnet" { 12u64 } else { 100u64 };
+        let mut epoch_length = 30_000;
+        let delay_blocks = {
+            if &network == "Mainnet" {
+                12u64
+            } else if network.starts_with("Heco") || network.starts_with("BSC") {
+                epoch_length = 200;
+                11u64
```
Comment (wuminzhe) [inline src/mmr/runner.rs:43]: why 11?

Comment (wuminzhe) [reply src/mmr/runner.rs:43]: I know

**Review by wuminzhe — APPROVED** [review]

**Comment by HackFisher** [comment]
@songtianyi Please fix the conflicts

**Review by HackFisher — CHANGES_REQUESTED** [review]
Following this implementation, so we are going to deploy separate shadow instance service for different networks? And the api format keeps the same, e.g. `ethereum/mmr_leaf`.

Though using the same api may cause confusion for api user, I'm ok with this approach for now. But please document the README more clearly, for easier understanding to deployer and api user.

**Comment by wuminzhe** [comment]
Maybe you can a network name or id to the resp or http header to help the client identify the network more easily.

File: README.md [hunk 3]
```diff
@@ -22,45 +21,60 @@ FLAGS:
...
+* `MMR_LOG`
+
+    Optional. Define how frequently it outputs logs `Pushed mmr ... into database` while generating MMR. Useful when you first time running shadow, since it generates millon of MMR data at first launch. Default is `10000` .
+
+    Example: `"100000"`
 
 ## Trouble Shooting
 
-Everytime you run `proof` in error, please delete `~/.ethashproof` and `~/.ethash` 
+Everytime you run `proof` in error, please delete `~/.ethashproof` and `~/.ethash`
```
Comment (xiaoch05) [inline README.md:76]: these two files have been moved to `~/.shadow/`

File: src/cmd/run.rs [hunk 4]
```diff
@@ -51,6 +56,14 @@ fn ethereum_rpc() -> EthereumRPC {
...
                 crate::conf::DEFAULT_ETHEREUM_ROPSTEN_RPC.into()
+            } else if env::var("HECO_MAINNET").is_ok() {
+                crate::conf::DEFAULT_HECO_MAINNET_RPC.into()
+            } else if env::var("HECO_TESTNET").is_ok() {
+                crate::conf::DEFAULT_HECO_TESTNET_RPC.into()
+            } else if env::var("BSC_MAINNET").is_ok() {
+                crate::conf::DEFAULT_BSC_MAINNET_RPC.into()
+            } else if env::var("BSC_TESTNET").is_ok() {
+                crate::conf::DEFAULT_BSC_TESTNET_RPC.into()
             } else {
```
Comment (xiaoch05) [inline src/cmd/run.rs:67]: for different networks, if we start separate shadow instances, why we don't use a single environment variable such as `ETHEREUM_RPC`.

Comment (songtianyi) [reply src/cmd/run.rs:67]: Yes, a single env variable is enough. i followed the old convention `ETHEREUM_RPC `, `ETHEREUM_ROPSTEN` to add these new envs

Comment (songtianyi) [reply src/cmd/run.rs:67]: When we start several different shadow instance in the same host, that would be necessary to distinguish different instance by envs

File: README.md (hunk 3 above)
Comment (songtianyi) [reply README.md:76]: ok

File: src/cmd/run.rs (hunk 4 above)
Comment (xiaoch05) [reply src/cmd/run.rs:67]: if the env is global. the `ETHEREUM_RPC` env will prevent the others. So it may be better to use a local env for each instance.

Comment (songtianyi) [reply src/cmd/run.rs:67]: I'm wrong about this i think. It's not possible to start different instance in same host by different envs

Comment (songtianyi) [reply src/cmd/run.rs:67]: Yes, that's the only way

**Comment by songtianyi** [comment]
> Maybe you can add a network name or id to the resp or http header to help the client identify the network more easily.

I'm not sure about this, the clients should know who they are accessing to

**Comment by HackFisher** [comment]
Keep it for now, we may do not need shadow mmr service for bsc is we put the message in storage, but the down side is that we will need to verify two layer merkle tree (storage root to account, account root to message)

**Comment by HackFisher** [comment]
Will use the graph to generate mmr instead if required.

**Pull request closed without merging**